### ISSUE TITLE
Thread count increases

### ISSUE THREAD

**Issue by winstonma** [agronholm/apscheduler#207]
Hi I have faced some trouble while using apscheduler.

I followed the [Blocking Scheduler Demo](https://github.com/agronholm/apscheduler/blob/master/examples/schedulers/blocking.py) and I added the thread count in the function `tick.` Here is my [gist](https://gist.github.com/winstonma/1cef3927ce716170966b05832a6bfae5).

When I run the program and here is the output
```
Tick! The time is: 2017-06-02 14:34:05.504860
Thread Count: 2
Tick! The time is: 2017-06-02 14:34:06.509417
Thread Count: 3
Tick! The time is: 2017-06-02 14:34:07.505808
Thread Count: 4
Tick! The time is: 2017-06-02 14:34:08.509687
Thread Count: 5
Tick! The time is: 2017-06-02 14:34:09.504899
Thread Count: 6
Tick! The time is: 2017-06-02 14:34:10.509752
Thread Count: 7
Tick! The time is: 2017-06-02 14:34:11.509402
Thread Count: 8
Tick! The time is: 2017-06-02 14:34:12.506539
Thread Count: 9
Tick! The time is: 2017-06-02 14:34:13.507271
Thread Count: 10
Tick! The time is: 2017-06-02 14:34:14.509713
Thread Count: 11
Tick! The time is: 2017-06-02 14:34:15.509802
Thread Count: 11
Tick! The time is: 2017-06-02 14:34:16.504057
Thread Count: 11
Tick! The time is: 2017-06-02 14:34:17.509134
Thread Count: 11
Tick! The time is: 2017-06-02 14:34:18.504486
Thread Count: 11
```

The thread count goes up when the function is being executed. I am not sure if that is expected. Thanks for taking a look.

**Comment by winstonma** [comment]
I also tried the [asyncio](https://github.com/agronholm/apscheduler/blob/master/examples/schedulers/asyncio_.py) version that would increase the thread count to 21.

**Comment by agronholm** [comment]
The standard library ThreadPoolExecutor is lazy and just adds a thread whenever a new task is received (up to the maximum worker count). So this is expected.

**Comment by winstonma** [comment]
Thanks

The problem I am facing is using the using the APScheudler with aiohttp. The thread count would goes from 11 to 231.

I created another [gist](https://gist.github.com/winstonma/33eba585f2d785637bbcd4fba1fd4470) just for demostartion.

In every schedule I collected the HTML from Google.com. As a comparison the program would run a schedule function in loop before using the scheduler.

For the loop part the function would be like this:
First run the program without the scheduler
```
Tick! The time is: 2017-06-02 16:22:46.651977
Fetch finished
Thread Count: 3
Tick! The time is: 2017-06-02 16:22:47.322014
Fetch finished
Thread Count: 5
Tick! The time is: 2017-06-02 16:22:47.935847
Fetch finished
Thread Count: 7
Tick! The time is: 2017-06-02 16:22:48.554457
Fetch finished
Thread Count: 9
Tick! The time is: 2017-06-02 16:22:49.166575
Fetch finished
Thread Count: 11
Tick! The time is: 2017-06-02 16:22:49.875966
Fetch finished
Thread Count: 13
Tick! The time is: 2017-06-02 16:22:50.476179
Fetch finished
Thread Count: 15
Tick! The time is: 2017-06-02 16:22:52.011429
Fetch finished
Thread Count: 17
Tick! The time is: 2017-06-02 16:22:52.643666
Fetch finished
Thread Count: 19
Tick! The time is: 2017-06-02 16:22:53.266230
Fetch finished
Thread Count: 21
Tick! The time is: 2017-06-02 16:22:53.885696
Fetch finished
Thread Count: 21
Tick! The time is: 2017-06-02 16:22:54.498701
Fetch finished
Thread Count: 21
Tick! The time is: 2017-06-02 16:22:55.114552
Fetch finished
Thread Count: 21
Tick! The time is: 2017-06-02 16:22:56.825084
Fetch finished
Thread Count: 21
```

It keeps the thread count to 21.

Then the second part I use the scheduler to execute the function. And here is the log (with the steady thread count)
```
Tick! The time is: 2017-06-02 16:19:32.517794
Fetch finished
Thread Count: 231
Tick! The time is: 2017-06-02 16:19:33.519795
Fetch finished
Thread Count: 231
Tick! The time is: 2017-06-02 16:19:34.518189
```

It seems the thread count grows exponentially. Even I decrease the schedule period to 1 minute would not help.

The problem I faced is running Heroku which has the thread limit of 256 (Free tier) which the problem can be seen after running scheduler 15 times. I know this scheduler works perfectly but I have no idea how to get the resource released.

Thanks you very much

**Comment by winstonma** [comment]
What I am trying to say is the scheduler's thread count increase exponentially if the schedule function have threads running. While as you can see in the for loop the thread count won't goes up that high. I tried another scheduler which uses the same scheduler function would give me a stable thread count of 43, which is very similar to the for loop example.

**Comment by srPuebla** [comment]
> What I am trying to say is the scheduler's thread count increase exponentially if the schedule function have threads running. While as you can see in the for loop the thread count won't goes up that high. I tried another scheduler which uses the same scheduler function would give me a stable thread count of 43, which is very similar to the for loop example.

4 years later... have you solve this situation?

**Comment by agronholm** [comment]
> 4 years later... have you solve this situation?

The design flaw in `concurrent.futures.ThreadPoolExecutor` has been fixed in Python 3.8.
